File: prettier-plugin-csharpier/src/Printer/Types/ForStatement.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, printSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { AttributeListNode } from "./AttributeList";
import { printVariableDeclaration, VariableDeclarationNode } from "./VariableDeclaration";

export interface ForStatementNode extends SyntaxTreeNode<"ForStatement"> {
    attributeLists: AttributeListNode[];
    forKeyword?: SyntaxToken;
    openParenToken?: SyntaxToken;
    declaration?: VariableDeclarationNode;
    initializers: SyntaxTreeNode[];
    firstSemicolonToken?: SyntaxToken;
    condition?: SyntaxTreeNode;
    secondSemicolonToken?: SyntaxToken;
    incrementors: SyntaxTreeNode[];
    closeParenToken?: SyntaxToken;
    statement?: SyntaxTreeNode;
}

export const printForStatement: PrintMethod<ForStatementNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [printPathSyntaxToken(path, "forKeyword"), " ("];
    if (node.declaration) {
        parts.push(
            path.call(innerPath => printVariableDeclaration(innerPath, options, print), "declaration"),
            "; ",
        );
    } else {
        parts.push(";");
    }
    if (node.condition) {
        parts.push(path.call(print, "condition"), "; ");
    } else {
        parts.push(";");
    }

    parts.push(join(", ", path.map(print, "incrementors")), ")");

    let statement = path.call(print, "statement");
    if (node.statement?.nodeType === "Block") {
        parts.push(statement);
    } else {
        parts.push(indent(concat([hardline, statement])));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/ForStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class ForStatement
{
    public static Doc Print(ForStatementSyntax node, FormattingContext context)
    {
        var docs = new List<Doc>
        {
            ExtraNewLines.Print(node),
            Doc.Group(
                Token.Print(node.ForKeyword, context),
                " ",
                Token.Print(node.OpenParenToken, context),
                Doc.Group(
                    Doc.Indent(
                        Doc.SoftLine,
                        Doc.Group(
                            node.Declaration != null
                                ? VariableDeclaration.Print(node.Declaration, context)
                                : Doc.Null,
                            SeparatedSyntaxList.Print(node.Initializers, Node.Print, " ", context),
                            Token.Print(node.FirstSemicolonToken, context)
                        ),
                        node.Condition != null
                            ? Doc.Concat(Doc.Line, Node.Print(node.Condition, context))
                            : Doc.Line,
                        Token.Print(node.SecondSemicolonToken, context),
                        Doc.Line,
                        Doc.Group(
                            Doc.Indent(
                                SeparatedSyntaxList.Print(
                                    node.Incrementors,
                                    Node.Print,
                                    Doc.Line,
                                    context
                                )
                            )
                        )
                    ),
                    Doc.SoftLine
                ),
                Token.Print(node.CloseParenToken, context)
            ),
            node.Statement switch
            {
                ForStatementSyntax => Doc.Group(Doc.HardLine, Node.Print(node.Statement, context)),
                _ => OptionalBraces.Print(node.Statement, context),
            },
        };

        return Doc.Concat(docs);
    }
}
